File: packages/shared/src/actions.ts
// The actions the user can trigger from the popover.
// Kept as a const-tuple + union so it works in both runtime (zod enum, switches)
// and compile-time (exhaustive checks).
//
//   reply     — draft a contextual response to a conversation
//   grammar   — fix grammar/spelling in the user's draft, in its own language
//   rewrite   — transform/compose/light-edit text (optionally into a new language)
//   translate — render a customer query (or any text) in a chosen language

export const ACTIONS = ["reply", "grammar", "rewrite", "translate"] as const;
export type Action = (typeof ACTIONS)[number];

export const isAction = (value: unknown): value is Action =>
  typeof value === "string" && (ACTIONS as readonly string[]).includes(value);

File: backend/src/inkwell_backend/domain/actions.py
"""Actions the user can trigger from the popover.

Mirrors `@inkwell/shared/actions`. `Action` is a string-valued enum so
Pydantic validates request payloads against the same set the TypeScript
backend does, and IDE autocomplete still works inside the codebase.
"""

from __future__ import annotations

from enum import StrEnum


class Action(StrEnum):
    """Every action surface the extension can request from the backend."""

    # Draft a contextual response to a conversation.
    REPLY = "reply"
    # Fix grammar/spelling in the user's draft, in its own language.
    GRAMMAR = "grammar"
    # Transform / compose / light-edit text (optionally into a new language).
    REWRITE = "rewrite"
    # Render a customer query (or any text) in a chosen language.
    TRANSLATE = "translate"
